lib/internal/streams/operators.js

```diff
@@ -0,0 +1,234 @@
+  }
+
+  const ac = new AbortController();
+  const abort = () => ac.abort();
+
+  const queue = [];
+  const signalOpt = { signal: ac.signal };
+  const flatten = !!config?.flatten;
+  const outerSignal = options?.signal;
+
+  outerSignal?.addEventListener('abort', abort);
+
+  let next;
+  let resume;
+  let done = false;
+
+  async function pump () {
+    try {
+      for await (let val of stream) {
+        if (done) {
+          return;
+        }
+
+        try {
+          val = fn(val, signalOpt);
+        } catch (err) {
+          val = Promise.reject(err);
+        }
+
+        if (val === kEmpty) {
+          continue;
+        }
+
+        if (typeof val?.then === 'function') {
+          val.then(null, () => {
+            done = true;
+          });
+        }
+
+        queue.push(val);
+        if (next) {
+          next();
+          next = null;
+        }
+
+        if (!done && queue.length >= concurrency) {
+          await new Promise(resolve => {
+            resume = resolve;
+          });
+        }
+      }
+      queue.push(null);
+    } catch (err) {
+      queue.push(Promise.reject(err));
+    } finally {
+      done = true;
+      if (next) {
+        next();
+        next = null;
+      }
+    }
+  }
+
+  pump();
+
+  try {
+    while (true) {
+      while (queue.length > 0) {
+        let val = queue[0];
+
+        if (val === null) {
+          return;
+        }
+
+        if (flatten && isIterable(val)) {
+          yield * val;
+        } else {
+          val = await val;
+          if (val !== kEmpty) {
+            yield val;
+          }
+        }
+
+        queue.shift();
+        if (resume) {
+          resume();
+          resume = null;
+        }
+      }
+
+      await new Promise(resolve => {
+        next = resolve;
+      });
+    }
+  } finally {
+    ac.abort();
+
+    done = true;
+    if (resume) {
+      resume();
+      resume = null;
+    }
+
+    outerSignal?.removeEventListener('abort', abort);
+  }
+}
+
+function map (fn, options) {
+  return mapImpl(this, fn, options);
+}
+
+function filter (fn, options) {
+  return map(this, async val => await fn(val) ? val : kEmpty, options);
+}
+
+function drop (n, options) {
+  return map(this, val => {
+    n -= 1;
+    return n >= 0 ? null : val;
+  }, options);
+}
+
+function take (n, options) {
+  return map(this, val => {
+    n -= 1;
+    return n >= 0 ? val : null;
+  }, options);
+}
+
+function asIndexedPairs (options) {
+  let n = 0;
+  return map(this, val => [n++, val], options);
+}
+
+function flatMap (fn, options) {
+  return mapImpl(this, fn, options, { flatten: true });
+}
+
+async function reduce (reducer, initialValue, options) {
+  let acc = initialValue;
+  return map(this, async val => {
+    acc = acc !== undefined ? await reducer(acc, val) : val;
+    return acc;
+  }, options);
+}
+
+async function toArray () {
+  const arr = [];
+  for await (const val of map(x => x, options)) {
```

Comment:
Any reason this goes through .map?

Reply:
concurrency

Reply:
Why would you want `toArray()` to be limited by `concurrency`?

Anyway I don't think it'd do anything but cycle the CPU since `x => x` finishes synchronously and `concurrency` is only meaningful for async actions.